core/domain/stats_services.py

```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        state_reference, entity_type)
+    if not learner_answer_details:
+        learner_answer_details = stats_domain.LearnerAnswerDetails(
+            state_reference, entity_type, interaction_id, [], 0)
+    learner_answer_info_id = (
+        stats_domain.LearnerAnswerInfo.get_learner_answer_info_id())
+    learner_answer_info = stats_domain.LearnerAnswerInfo(
+        learner_answer_info_id, answer, answer_details,
+        datetime.datetime.utcnow())
+    learner_answer_details.add_learner_answer_info(learner_answer_info)
+    save_learner_answer_details(
+        learner_answer_details, state_reference, entity_type)
+
+
+def delete_learner_answer_info(
+        state_reference, entity_type, learner_answer_info_id):
+    """Deletes the learner answer info in the model, and then
+    saves it.
+
+    Args:
+        state_reference: str. This is used to refer to a state
+            in an exploration or question. For an exploration the
+            value will be equal to 'exp_id.state_name' and for question
+            this will be equal to 'question_id'.
+        entity_type: str. The type of entity i.e ENTITY_TYPE_EXPLORATION
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+        learner_answer_info_id: str. The unique ID of the learner answer info
+            which needs to be deleted.
+"""
+    learner_answer_details = get_learner_answer_details(
+        state_reference, entity_type)
+    if not learner_answer_details:
+        raise utils.InvalidInputException(
+            'No learner answer details found with the given state '
+            'reference and entity')
+    learner_answer_details.delete_learner_answer_info(
+        learner_answer_info_id)
+    save_learner_answer_details(
+        learner_answer_details, state_reference, entity_type)
+
+
+def update_state_reference(
+        new_state_reference, old_state_reference, entity_type):
+    """Updates the state reference field of the model instance if the
+    and then saves it.
+
+    Args:
+        new_state_reference: str. The new state reference which needs to be
+            updated..
+        old_state_reference: str. The old state reference which needs to be
+            changed.
+        entity_type: str. The type of entity i.e ENTITY_TYPE_EXPLORATION
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+    """
+    learner_answer_details = get_learner_answer_details(
+        old_state_reference, entity_type)
+    if not learner_answer_details:
+        raise utils.InvalidInputException(
+            'No learner answer details found with the given state '
+            'reference and entity')
+    learner_answer_details.update_state_reference(new_state_reference)
+    save_learner_answer_details(
+        learner_answer_details, old_state_reference, entity_type)
+
+
+def delete_learner_answer_details_for_exploration_state(
+        exp_id, state_name):
+    """Delete the LearnerAnswerDetailsModel instance of entity type
```

Comment:
You can just say: Deletes the LearnerAnswerDetailsModel corresponding to the given exploration ID and state name.

Reply:
Done